File: frontend/src/lib/api/ratings.ts
import { apiClient } from './client';

export interface RatingCreate {
  transaction_type: 'PRODUCT_PURCHASE' | 'DEMAND_FULFILLMENT';
  reference_id: string;
  score: number;
  comment?: string;
}

export interface RatingResponse {
  id: string;
  rater_id: string;
  rated_id: string;
  role_context: 'AS_SELLER' | 'AS_BUYER';
  transaction_type: 'PRODUCT_PURCHASE' | 'DEMAND_FULFILLMENT';
  reference_id: string;
  score: number;
  comment?: string;
  created_at: string;
  rater_name?: string;
}

export const ratingsApi = {
  submitRating: async (data: RatingCreate) => {
    const response = await apiClient('/ratings', {
      method: 'POST',
      body: JSON.stringify(data),
    });
    return response.json();
  },

  getUserRatingsAsSeller: async (userId: string) => {
    const response = await apiClient(`/ratings/user/${userId}/seller`, {
      method: 'GET',
    });
    return response.json();
  },

  getUserRatingsAsBuyer: async (userId: string) => {
    const response = await apiClient(`/ratings/user/${userId}/buyer`, {
      method: 'GET',
    });
    return response.json();
  },
};

File: backend/app/routers/ratings.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from uuid import UUID
from typing import List, Optional
from datetime import datetime

from app.db import get_db
from app.models.user import User
from app.models.rating import Rating, RoleContext, TransactionType
from app.services import auth_service
from app.services.rating_service import create_rating, get_user_ratings

from app.schemas.rating import RatingCreateSchema, RatingResponseSchema, RatingSummarySchema

router = APIRouter(prefix="/ratings", tags=["ratings"])

@router.post("", response_model=RatingResponseSchema)
async def submit_rating_endpoint(
    body: RatingCreateSchema,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(auth_service.get_current_user)
):
    rating = await create_rating(
        db=db,
        rater_id=current_user.id,
        transaction_type=body.transaction_type,
        reference_id=body.reference_id,
        score=body.score,
        comment=body.comment
    )
    return rating

@router.get("/user/{user_id}/seller", response_model=RatingSummarySchema)
async def get_user_ratings_as_seller(
    user_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    return await get_user_ratings(db, user_id, RoleContext.AS_SELLER)

@router.get("/user/{user_id}/buyer", response_model=RatingSummarySchema)
async def get_user_ratings_as_buyer(
    user_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    return await get_user_ratings(db, user_id, RoleContext.AS_BUYER)
